### PR TITLE
Default to external cloud provider

### LINKED ISSUE #188: Default template should use out-of-tree cloud provider
Now that CAPZ has switched to using the recommended OOT cloud-provider-azure component, `az capi` should do the same. We should also ensure that user-supplied templates get the OOT componentry installed if needed.

### PR DESCRIPTION
**Pull request by mboersma**
**Description**

Sets `--external-cloud-provider=true` as the default to `az capi create`. It's time; without this, Kubernetes v1.29.x clusters won't work.

Fixes #188

**History Notes**

`--external-cloud-provider=true` is now the default for `az capi create`

---

This checklist is used to make sure that common guidelines for an Azure CLI pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).
- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).
- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/capi/azext_capi/custom.py [hunk 1]
```diff
@@ -469,8 +468,8 @@ def create_workload_cluster(  # pylint: disable=too-many-arguments,too-many-loca
         defined_args = [v["name"] for v in mutual_exclusive_args if v["value"]]
         if defined_args:
             defined_args = " ,".join(defined_args)
-            error_msg = f'The following arguments are incompatible with "--template":\n{defined_args}'
-            raise MutuallyExclusiveArgumentError(error_msg)
+            warning_msg = f'The following arguments are incompatible with "--template":\n{defined_args}'
+            logger.warning(warning_msg)
```
Comment (willie-yao) [inline src/capi/azext_capi/custom.py:472]: Why is this change needed? Did the function use to terminate after the error?

File: src/capi/azext_capi/custom.py [hunk 2]
```diff
@@ -597,7 +597,28 @@ def install_cni(cmd, cluster_name, workload_cfg, windows, args):
             cidr1 = retry_shell_command(command, attempts, delay)
         except subprocess.CalledProcessError:
             pass  # This is a best-effort configuration, so don't fail if we can't find the CIDR.
+    return cidr0, cidr1
+
+
+def install_cloud_provider(cmd, cluster_name, workload_cfg, cidr0, cidr1):
```
Comment (willie-yao) [inline src/capi/azext_capi/custom.py:603]: What is cidr0 and cidr1? I'm not too familiar with the terms and it may help to add a comment explaining the parameters.

**Review by willie-yao — COMMENTED** [review]
Just a few questions from my end! I was wondering why there were a few changes in the logic if we're changing the default value.

File: src/capi/azext_capi/custom.py (hunk 1 above)
Comment (mboersma) [reply src/capi/azext_capi/custom.py:472]: Yes, sorry this is a logic change that I should have explained up front. This used to be a hard error, but in testing the external-cloud-provider changes in the pipeline I realized we use a local template there but now want to set the `--external-cloud-provider` flag. 

So I decided to make all these cases a yellow warning message rather than fail. It's basically saying if your user-supplied template doesn't have sections like 

```jinja2
{% if EXTERNAL_CLOUD_PROVIDER %}external{% else %}azure{% endif %}
```

then it's not likely to work with those flags.

File: src/capi/azext_capi/custom.py (hunk 2 above)
Comment (mboersma) [reply src/capi/azext_capi/custom.py:603]: [CIDR](https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing) == classless inter-domain routing, here meaning basically "What network address ranges are available to this cluster?" Definitely should add a comment; I'll do that.

Comment (mboersma) [reply src/capi/azext_capi/custom.py:607]: Added some comments about CIDRs and also an explanation for the "incompatible flags" warning.

File: src/capi/azext_capi/custom.py [hunk 3]
```diff
@@ -520,17 +521,12 @@ def create_workload_cluster(  # pylint: disable=too-many-arguments,too-many-loca
         "PRE_BOOTSTRAP_CMDS": bootstrap_cmds["pre"],
         "POST_BOOTSTRAP_CMDS": bootstrap_cmds["post"],
         "AKS_INFRA_RG_NAME": os.environ.get(AKS_INFRA_RG_NAME, None),
-        "AKS_VNET_NAME": os.environ.get(AKS_VNET_NAME, None)
+        "AKS_VNET_NAME": os.environ.get(AKS_VNET_NAME, None),
+        "EXTERNAL_CLOUD_PROVIDER": external_cloud_provider,
```
Comment (mboersma) [inline src/capi/azext_capi/custom.py:525]: This change is paired with the previous one that doesn't prevent the user from adding these flags/variables if they know what they're doing.

File: src/capi/azext_capi/custom.py (hunk 2 above)
Comment (willie-yao) [reply src/capi/azext_capi/custom.py:607]: Ah I was confused about the difference/meaning of cidr0 vs cidr1 but the comment is helpful. Thanks!

**Review by willie-yao — APPROVED** [review]
/lgtm

File: src/capi/azext_capi/custom.py (hunk 2 above)
Comment (mboersma) [reply src/capi/azext_capi/custom.py:607]: Oh if you were curious about the variable naming, it's just "first CIDR" and "second CIDR" but zero-based, I think this was copied directly from some CAPZ shell script of yore.
